Automattic/Edit-Flow
# Keep save button text updated, without errors.

Fixes https://github.com/Automattic/Edit-Flow/issues/583. A different approach to https://github.com/Automattic/Edit-Flow/pull/584.

### Fixing the page-breaking error
On the initial load, when the `subscribe()` method triggers we can update the button text right away from when it first exists. By checking specifically if the `innerText` is one of the two values we don't want, we avoid the errors that occur in https://github.com/Automattic/Edit-Flow/issues/583.

### Keeping the button text changes even after post saving
Then there is the issue of the button text reverting back to "Save draft" / "Save as Pending" after saving the post.

When a post is being saved, the DOM change quite a bit (4 different states) and there is unfortunately no way that I could find to hook in after the save is completed and the button is back to it's default state (the last animation tick being particularly annoying). To workaround this limitation, I'm using the MutationObserver API if the browser has it available: https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver

So then when the `subscribe()` method is triggered and the post is being saved, we set up a single observer on the parent element. This observer waits for the button to be added back then it targets the `innerText` again and get it updated. This node is actually re-added quite a few times per post-save, and due to various animations it's not actually possible to set/remove the observer each time. So instead, I decided to just set up one observer and keep it running. It watches over the minimal amount of nodes necessary.

### Testing
1. Install and activate Gutenberg plugin (7.3)
2. Ensure custom statuses module is enabled.
3. Create/edit posts, change around statuses, re-open old drafts, etc.
4. Make sure the button always just says "Save" rather than "Save Draft", etc.
5. Do the same tests without the GB plugin active, for BC testing.
